indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java

```diff
@@ -1717,10 +1738,161 @@ public Response getUnparseableEvents(
+
+    if (lastSequenceMetadata.startOffsets.equals(offsetsForLastPartitionAssignment)) {
+      // This is the case where no data has been ingested since the last successfull checkpoint for these partitions.
+      // In this case, we'll mark the end offsets for this sequence same as start offsets and create a new sequence.
+      log.info("No new data ingested across any sibling tasks for this partition sequence since last checkpoint.");
+      lastSequenceMetadata.setEndOffsets(lastSequenceMetadata.startOffsets);
+    }
+
+    final Set<PartitionIdType> exclusiveStartPartitions = computeExclusiveStartPartitionsForSequence(
+        partitionStartOffsets);
+
+    final SequenceMetadata<PartitionIdType, SequenceOffsetType> newSequence = new SequenceMetadata<>(
+        sequences.isEmpty() ? 0 : lastSequenceMetadata.getSequenceId() + 1,
+        StringUtils.format(
+            "%s_%d",
+            ioConfig.getBaseSequenceName(),
+            sequences.isEmpty() ? 0 : lastSequenceMetadata.getSequenceId() + 1
+        ),
+        partitionStartOffsets,
+        partitionEndSequences,
+        false,
+        exclusiveStartPartitions,
+        getTaskLockType()
+    );
+    log.info("Attempting adding new sequence [%s]", newSequence);
+
+    currOffsets.clear();
+    currOffsets.putAll(partitionStartOffsets);
+    endOffsets.clear();
+    endOffsets.putAll(partitionEndSequences);
+
+    addSequence(newSequence);
+    persistSequences();
+    log.info(
+        "Created new sequence [%s] with start offsets [%s]",
+        newSequence.getSequenceName(), partitionStartOffsets
+    );
+  }
+
+  private void checkpointSequences()
+  {
+    try {
+      final SequenceMetadata<PartitionIdType, SequenceOffsetType> latestSequence = getLastSequenceMetadata();
+      if (!latestSequence.isCheckpointed()) {
+        final CheckPointDataSourceMetadataAction checkpointAction = new CheckPointDataSourceMetadataAction(
+            getSupervisorId(),
+            ioConfig.getTaskGroupId(),
+            null,
+            createDataSourceMetadata(
+                new SeekableStreamStartSequenceNumbers<>(
+                    stream,
+                    latestSequence.getStartOffsets(),
+                    latestSequence.getExclusiveStartPartitions()
+                )
+            )
+        );
+        if (!toolbox.getTaskActionClient().submit(checkpointAction)) {
+          throw new ISE("Checkpoint request with sequences [%s] failed, dying", currOffsets);
+        }
+      }
+    }
+    catch (Exception e) {
+      log.error(e, "Failed to checkpoint sequences.");
+      backgroundThreadException = e;
+    }
+  }
+
   @VisibleForTesting
   public Response setEndOffsets(
       Map<PartitionIdType, SequenceOffsetType> sequenceNumbers,
-      boolean finish // this field is only for internal purposes, shouldn't be usually set by users
+      boolean finish, // this field is only for internal purposes, shouldn't be usually set by users
```

Comment:
This old comment does not add any value. Please remove it.

```suggestion
      boolean finish,
```